### PR TITLE
feat: storage-cost per storage-provider

### PR DESCRIPTION
**Pull request by shifty11**
## Summary by CodeRabbit

- **New Features**
  - Enhanced storage cost management with dynamic pricing per storage provider.
  - Added migration logic for storage costs to support upgrades and ensure data consistency.
  - Updated API documentation to reflect changes in storage cost structure.

- **Documentation**
  - Updated OpenAPI specifications to align with the new storage cost management system.

- **Bug Fixes**
  - Adjusted storage cost retrieval logic across various test scenarios to align with the new dynamic pricing model.

- **Refactor**
  - Revamped parameter handling in the `x/bundles` module to utilize a map structure for storage costs, enhancing flexibility and scalability.

### CODE REVIEW

File: app/upgrades/v1_5/v1_4_types/getters_params.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// GetParams returns the current x/bundles module parameters.
+func GetParams(ctx sdk.Context, storeKey storeTypes.StoreKey, cdc codec.Codec) (params Params) {
+	store := ctx.KVStore(storeKey)
+
+	bz := store.Get(types.ParamsPrefix)
+	if bz == nil {
+		return params
+	}
+
+	cdc.MustUnmarshal(bz, &params)
```
[the remark this answers is not in the document]
Comment (shifty11) [reply app/upgrades/v1_5/v1_4_types/getters_params.go:19]: No we don't do that

File: proto/kyve/bundles/v1beta1/params.proto [hunk 2]
```diff
@@ -6,15 +6,21 @@ import "gogoproto/gogo.proto";
...
 // Params defines the bundles module parameters.
 message Params {
   // upload_timeout ...
   uint64 upload_timeout = 1;
   // storage_cost ...
-  string storage_cost = 2 [
-    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
-    (gogoproto.nullable) = false
-  ];
+  map<uint32, LegacyDecValue> storage_costs = 2 [(gogoproto.nullable) = false];
```
Comment (mbreithecker) [inline proto/kyve/bundles/v1beta1/params.proto]: The map iterator is not deterministic and therefore can not be used

**Review by mbreithecker — CHANGES_REQUESTED** [review]
A Go map can not be used for proto marshalling because the iterator is not deterministic. This needs to be changed to an array.

I would then recommend to just build a get and set function which just iterates the array.
Because it's marshalled into one proto object and always gets loaded and stored as a whole, the O(n) for the small in memory list iteration is negligible.

File: proto/kyve/bundles/v1beta1/params.proto (hunk 2 above)
Comment (shifty11) [reply proto/kyve/bundles/v1beta1/params.proto]: It is now an array. 
My assumption is that new storage provider ID's are always incremented by 1, so I can access the array by index with the storage id.

File: app/upgrades/v1_5/upgrade.go [hunk 3]
```diff
@@ -14,11 +21,57 @@ const (
...
+	oldParams := v1_4_types.GetParams(sdkCtx, bundlesStoreKey, cdc)
+	newParams := bundlestypes.Params{
+		UploadTimeout: oldParams.UploadTimeout,
+		StorageCosts:  make([]math.LegacyDec, len(storageIds)),
+		NetworkFee:    oldParams.NetworkFee,
+		MaxPoints:     oldParams.MaxPoints,
+	}
+	for storageId := range storageIds {
+		if int(storageId) > len(newParams.StorageCosts) {
+			return fmt.Errorf("storage provider id %d is out of bounds", storageId)
```
Comment (mbreithecker) [inline app/upgrades/v1_5/upgrade.go]: remove this when we turn the array into an array of {id, storageCost}

Also lets then just hardcode the value for Storage Id 1 and 2. 
A few days before finalization we will calculate the current values and use those

File: proto/kyve/bundles/v1beta1/params.proto [hunk 4]
```diff
@@ -10,8 +10,8 @@ option go_package = "github.com/KYVENetwork/chain/x/bundles/types";
 message Params {
   // upload_timeout ...
   uint64 upload_timeout = 1;
-  // storage_cost ...
-  string storage_cost = 2 [
+  // storage_costs ...
+  repeated string storage_costs = 2 [
```
Comment (mbreithecker) [inline proto/kyve/bundles/v1beta1/params.proto]: Create a separate struct with 
message StorageCostEntry {
  uint64 id = 1;
  (legacyDec...) storage_cost = 2 [...];
}

File: x/bundles/keeper/getters_params.go [hunk 5]
```diff
@@ -26,8 +21,12 @@ func (k Keeper) GetUploadTimeout(ctx sdk.Context) (res uint64) {
 }
 
 // GetStorageCost returns the StorageCost param
-func (k Keeper) GetStorageCost(ctx sdk.Context) (res math.LegacyDec) {
-	return k.GetParams(ctx).StorageCost
+func (k Keeper) GetStorageCost(ctx sdk.Context, storageProviderId uint32) (res math.LegacyDec) {
+	storageCosts := k.GetParams(ctx).StorageCosts
+	if int(storageProviderId) < len(storageCosts) {
```
Comment (mbreithecker) [inline x/bundles/keeper/getters_params.go]: Iterate the array until id = storageProviderId
if not match => return 0

File: x/bundles/keeper/msg_server_update_params_test.go [hunk 6]
```diff
@@ -67,7 +67,7 @@ var _ = Describe("msg_server_update_params.go", Ordered, func() {
 		params := s.App().BundlesKeeper.GetParams(s.Ctx())
 
 		Expect(params.UploadTimeout).To(Equal(types.DefaultUploadTimeout))
-		Expect(params.StorageCost).To(Equal(types.DefaultStorageCost))
+		Expect(params.StorageCosts).To(Equal(types.DefaultStorageCosts()))
```
Comment (mbreithecker) [inline x/bundles/keeper/msg_server_update_params_test.go]: DefaultStorageCosts will be empty array

File: x/bundles/keeper/msg_server_update_params_test.go [hunk 7]
```diff
@@ -108,7 +108,9 @@ var _ = Describe("msg_server_update_params.go", Ordered, func() {
 		// ARRANGE
 		payload := `{
 			"upload_timeout": 20,
-			"storage_cost": "0.050000000000000000",
+			"storage_costs": [
+				"0.05"
```
Comment (mbreithecker) [inline x/bundles/keeper/msg_server_update_params_test.go]: refactor to array of {id, storageCost}

**Review by mbreithecker — CHANGES_REQUESTED** [review]

**Review by mbreithecker — APPROVED** [review]

**Pull request merged**